lib/declare_schema/model/habtm_model_shim.rb

```diff
@@ -5,28 +5,21 @@ module Model
     class HabtmModelShim
       class << self
         def from_reflection(refl)
-          join_table = refl.join_table
-          foreign_keys_and_classes = [
-            [refl.foreign_key.to_s, refl.active_record],
-            [refl.association_foreign_key.to_s, refl.class_name.constantize]
-          ].sort { |a, b| a.first <=> b.first }
-          foreign_keys = foreign_keys_and_classes.map(&:first)
-          foreign_key_classes = foreign_keys_and_classes.map(&:last)
-          # this may fail in weird ways if HABTM is running across two DB connections (assuming that's even supported)
-          # figure that anybody who sets THAT up can deal with their own migrations...
-          connection = refl.active_record.connection
-
-          new(join_table, foreign_keys, foreign_key_classes, connection)
+          new(refl.join_table, [refl.foreign_key, refl.association_foreign_key],
+                               [refl.active_record.table_name, refl.class_name.constantize.table_name])
         end
       end
 
-      attr_reader :join_table, :foreign_keys, :foreign_key_classes, :connection
+      attr_reader :join_table, :foreign_keys, :table_names
 
-      def initialize(join_table, foreign_keys, foreign_key_classes, connection)
+      def initialize(join_table, foreign_keys, table_names)
+        foreign_keys.is_a?(Array) && foreign_keys.size == 2 or
+          raise ArgumentError, "foreign_keys must be <Array[2]>; got #{foreign_keys.inspect}"
+        table_names.is_a?(Array) && table_names.size == 2 or
+          raise ArgumentError, "table_names must be <Array[2]>; got #{table_names.inspect}"
         @join_table = join_table
-        @foreign_keys = foreign_keys
-        @foreign_key_classes = foreign_key_classes
-        @connection = connection
+        @foreign_keys = foreign_keys.sort
+        @table_names = @foreign_keys == foreign_keys ? table_names : table_names.reverse
```

Comment:
I think it's a little confusing for there to be both a table_name and table_names defined since they aren't related to each other. Maybe this could be `parent_table_names` or `foreign_key_table_names` to be explicit about the difference?

Reply:
Good call. Updated here: https://github.com/Invoca/declare_schema/pull/157/commits/0c6cb41f88c5f30eb1fe84ad40e52a6418f913cb